Add Dream_oidc

andreypopp opened the pull request:
This adds `Dream_oidc` as a separate package.

andreypopp committed:
mv src/ oauth2/

andreypopp committed:
Hyper_helper now accepts Uri.t

Also handles adding Host header and 443 port for HTTPS automatically.

andreypopp committed:
Expose Internal.Hyper_helper

andreypopp committed:
Pin hyper to andreypopp/hyper

andreypopp committed:
Add dream-oidc package

aantron commented on oidc/dream_oidc.ml:
```diff
@@ -0,0 +1,214 @@
+
+(* Extract and validate id_token. *)
+let id_token config token =
+  (* TODO: this is from https://github.com/ulrikstrid/ocaml-oidc/pull/8 *)
```
I assume this is an adaptation rather than an exact copy. Nonetheless, how close is it? If it's very close, we probably need an explicit citation and acknowledgement in the README, and to preserve the [BSD license notice](https://github.com/ulrikstrid/ocaml-oidc/blob/2f8c4220098b0f1f97499331e2f9177b19f24ea1/LICENSE) either applied to only this file and called out in the README, or by changing this whole project to the BSD license and importing Ulrik's copyright notice from it.

This doesn't have to be done immediately, but I want to note it. If this code is that similar, the above has to be done before this project is released or advertised.

andreypopp replied:
Yes, hence the TODO

andreypopp committed:
Try more OIDC providers

andreypopp committed:
Update User_profile.t

- `display_name : string` -> `name : string option` (according to OIDC)
- add `email_verified : bool option` (according to OIDC)

andreypopp committed:
Allow to specify additional OIDC scope values

andreypopp committed:
Add preconfigured google/microsoft OIDC clients

andreypopp committed:
Convert Json_error exc to Error case

andreypopp committed:
Add nonce

The `nonce` param is used to prevent replay attacks. The possible attack
scenario:

- An attacker has access to Dream app to be able to get a fresh `state`
  CSRF token.

- An attacker observes a redirect back to OIDC client (dream app) from
  OIDC provider which has authorization `code` in its response.

- An attacker replays the same request to the OIDC client (dream app)
  with acquired `code` and fresh `state` CSRF token and triggers OIDC
  client to make another call to get another access token / id tokens
  and thus be able to auth as legitimate user.

Now if we use have `nonce` (which is bound to legitimate's user session)
in token response then OIDC client (dream app) will not perform
authentication of the attacker (as they don't have access to the
legitimate's user session).

Note that this implementation only performs a check if `nonce` received
from the OIDC provider is bound to the current request session.

The stronger protection could be achieved by making `nonce` (or `state`)
to be **truly nonce** (persist it outside of user agent and invalidate
on first use). In this case even if attacker has access to browser
session the replay won't be possible.

andreypopp commented:
Added `nonce` handling.

The `nonce` param is used to prevent replay attacks. The possible attack scenario:

- An attacker has access to Dream app to be able to get a fresh `state` CSRF token.

- An attacker observes a redirect back to OIDC client (dream app) from OIDC provider which has authorization `code` in its response.

- An attacker replays the same request to the OIDC client (dream app) with acquired `code` and fresh `state` CSRF token and triggers OIDC client to make another call to get another access token / id tokens and thus be able to auth as legitimate user.

Now if we have supplied `nonce` (which is bound to legitimate's user session) and it got back in token response then OIDC client (dream app) will not perform authentication of the attacker (as they don't have access to the legitimate's user session).

Note that this implementation only performs a check if `nonce` received from the
OIDC provider is bound to the current request session.

I believe this is enough protection against an attack described above (where attackes doesn't have access to browser session).

The stronger protection could be achieved by making `nonce` (or `state`) to be **truly nonce** (persist it outside of user agent and invalidate on first use, this means `Dream.cookie_sessions` is not usable here). In this case even if attacker has access to browser session the replay won't be possible.

andreypopp committed:
Update code for recent changes in oidc

andreypopp committed:
Allow to specify user claims in OIDC

andreypopp committed:
Add User_profile.json with orig json payload

This is useful if one wants to extract some specific info from userinfo
response. Also this is essential if we allow scopes/claims to be
configurable.

andreypopp committed:
Separate API for configuring OIDC clients

andreypopp committed:
Make OAuth2 clients types abstract

andreypopp committed:
Minor cleanup
